### ISSUE TITLE
Semaphore deadlock on nonblocking acquiring

### ISSUE THREAD

**Issue by anskornyakov** [python-zk/kazoo#429]
I have a bunch of services running thousands instances which using synchronization via kazoo zookeeper semaphore implementation. Network issues on particular instance or net segment is an everyday occurrence and I m facing deadlocks on my semaphores hardly when such issues appears.

The reason of a deadlock is an zombie lock nodes, it typically looks like:
```
(CONNECTED) /some_node/shard-semaphore-__lock__> ls
ecf828622e7d432bb5801267456599fc__lock__0000000001    // zombie lock node
ccf828622e7d432bb5801267456599fc__lock__0000000191
acf828622e7d432bb5801267456599fc__lock__0000000192
dcf828622e7d432bb5801267a56599fc__lock__0000000194
```

Usually zombie node owner is an instance that failed to delete that node due to network issues the go through and trying to acquire semaphore again and creating new lock node, but as soon as zombie node with  less sequence number exists, it or another instance will never be succeed in obtaining semaphore, and that's why here we has a deadlock.

How does it happens? I'm using nonblocking acquiring.  So` blocking=False `here

 ```
#kazoo.recipe.lock.Lock
def acquire(self, blocking=True, timeout=None):
        def _acquire_lock():
            got_it = self._lock.acquire(False)
            if not got_it:
                raise ForceRetryError()
            return True

        retry = self._retry.copy()
        retry.deadline = timeout

        # Ensure we are locked so that we avoid multiple threads in
        # this acquistion routine at the same time...
        locked = self._lock.acquire(False)
        if not locked and not blocking:
            return False
        if not locked:
            # Lock acquire doesn't take a timeout, so simulate it...
            try:
                locked = retry(_acquire_lock)
            except RetryFailedError:
                return False
        already_acquired = self.is_acquired
        try:
            gotten = False
            try:
                gotten = retry(self._inner_acquire,
                               blocking=blocking, timeout=timeout)  # <<<< 1. Assuming we are not blocking and failed to obtain lock, we get False here
            except RetryFailedError:
                pass
            except KazooException:
                # if we did ultimately fail, attempt to clean up
                exc_info = sys.exc_info()
                if not already_acquired:
                    self._best_effort_cleanup()
                    self.cancelled = False
                six.reraise(exc_info[0], exc_info[1], exc_info[2])
            if gotten:
                self.is_acquired = gotten
            if not gotten and not already_acquired:  # <<<< 2. Trying to do best effort cleanup
                self._best_effort_cleanup()
            return gotten    # <<<< 4. Return False but self.node is still alive
        finally:
            self._lock.release()

def _best_effort_cleanup(self):
    try:
        node = self.node or self._find_node()
        if node:
            self._delete_node(node)  # <<<< 3. Trying to delete node, but fail due to network issue, got and expcetion here but just swallow it and go through
    except KazooException:  # pragma: nocover
        pass

# kazoo.recipe.lock.Semaphore 
def _inner_acquire(self, blocking, timeout=None):
    self._session_expired = False
    self.client.add_listener(self._watch_session)

    if not self.assured_path:
        self._ensure_path()

    # Do we already have a lease?
    if self.client.exists(self.create_path):
        return True

    w = _Watch(duration=timeout)
    w.start()
    lock = self.client.Lock(self.lock_path, self.data)
    gotten = lock.acquire(blocking=blocking, timeout=w.leftover())  # <<<< 0. Trying to acquire lock nonblocking, as result got False and undeleted lock node here
    if not gotten:
        return False   <<<< 5. Our last chance lock.release() in finally block doesn't called, so lock node is not deleted
    try:
        while True:
            self.wake_event.clear()
            # Attempt to grab our lease...
            if self._get_lease():
                return True

            if blocking:
                # If blocking, wait until self._watch_lease_change() is
                # called before returning
                self.wake_event.wait(w.leftover())
                if not self.wake_event.isSet():
                    raise LockTimeout(
                        "Failed to acquire semaphore on %s "
                        "after %s seconds" % (self.path, timeout))
            else:
                return False
    finally:
        lock.release()
```

Here is zookeper log for deadlock case:
**>>>>> Semaphore ensures path <<<<<**
> 2017-04-16 21:51:38,732 DEBUG Sending request(xid=15): Exists(path=u/some_node/shard-semaphore, watcher=None)
> 2017-04-16 21:51:41,556 DEBUG Received response(xid=15): ZnodeStat(czxid=21477873603, mzxid=21477873936, ctime=1492368647519, mtime=1492368648479, version=1, cversion=1, aversion=0, ephemeralOwner=0, dataLength=1, numChildren=1, pzxid=21477877192)
> 2017-04-16 21:51:41,560 DEBUG Sending request(xid=16): GetData(path=u/some_node/shard-semaphore, watcher=None)
> 2017-04-16 21:51:44,145 DEBUG Received response(xid=16): (1, ZnodeStat(czxid=21477873603, mzxid=21477873936, ctime=1492368647519, mtime=1492368648479, version=1, cversion=1, aversion=0, ephemeralOwner=0, dataLength=1, numChildren=1, pzxid=21477877192))

**>>>>> Semaphore checks if we have lease  <<<<<**
> 2017-04-16 21:51:44,145 DEBUG Sending request(xid=17): Exists(path=u/some_node/shard-semaphore/9e5849446fbe4fa18a2e18b7098c9a95, watcher=None)

**>>>>> Lock ensures path  <<<<<**
> 2017-04-16 21:51:46,738 DEBUG Sending request(xid=18): Exists(path=u/some_node/shard-semaphore-__lock__, watcher=None)
> 2017-04-16 21:51:49,232 DEBUG Received response(xid=18): ZnodeStat(czxid=21477875405, mzxid=21477875405, ctime=1492368653082, mtime=1492368653082, version=0, cversion=5, aversion=0, ephemeralOwner=0, dataLength=0, numChildren=1, pzxid=21477879384)

**>>>>> Lock creates sequence ephemeral node  <<<<<**
> 2017-04-16 21:51:49,233 DEBUG Sending request(xid=19): Create(path=u/some_node/shard-semaphore-__lock__/e9eddb1df31949a59b5b7a49312367c4__lock__, data=xxx.yyy.net:28137, acl=[ACL(perms=31, acl_list=[ALL], id=Id(scheme=world, id=anyone))], flags=3)
> 2017-04-16 21:51:51,842 DEBUG Received response(xid=19): u/some_node/shard-semaphore-__lock__/e9eddb1df31949a59b5b7a49312367c4__lock__0000000003

**>>>>> Lock gets children, sort them and check if his node has least sequence number <<<<<**
> 2017-04-16 21:51:51,842 DEBUG Sending request(xid=20): GetChildren(path=u/some_node/shard-semaphore-__lock__, watcher=None)
> 2017-04-16 21:51:54,196 DEBUG Received response(xid=20): [ud981dca6219046498f544c052770652b__lock__0000000002, ue9eddb1df31949a59b5b7a49312367c4__lock__0000000003]

**>>>>> Lock acquiring failed, because other lock has least sequence number node, do best effort cleanup <<<<<**
> 2017-04-16 21:51:54,197 DEBUG Sending request(xid=21): Delete(path=u/some_node/shard-semaphore-__lock__/e9eddb1df31949a59b5b7a49312367c4__lock__0000000003, version=-1)

**>>>>> Delete failed due to network issues <<<<<**
> 2017-04-16 21:52:12,086 WARNING Connection dropped: socket connection broken
> 2017-04-16 21:52:12,086 WARNING Transition to CONNECTING
> 2017-04-16 21:52:12,086 INFO Zookeeper connection lost
> 2017-04-16 21:52:12,087 INFO Zookeeper connection suspended, pause actions.
> 2017-04-16 21:52:12,087 INFO Connecting to myzookeeper1.xxx.net:10650
> 2017-04-16 21:52:12,108 DEBUG Sending request(xid=None): Connect(protocol_version=0, last_zxid_seen=21477893974, time_out=60000, session_id=602194545035271393, passwd=Ojxeexx19xafx8bxbex08Yxf5/^oNx08, read_only=None)
> 2017-04-16 21:52:12,142 WARNING Connection dropped: socket connection broken
> 2017-04-16 21:52:12,142 INFO Connecting to myzookeeper4.xxx.net:10650
> 2017-04-16 21:52:12,157 DEBUG Sending request(xid=None): Connect(protocol_version=0, last_zxid_seen=21477893974, time_out=60000, session_id=602194545035271393, passwd=Ojxeexx19xafx8bxbex08Yxf5/^oNx08, read_only=None)
> 2017-04-16 21:52:12,289 INFO Zookeeper connection established, state: CONNECTED
> 2017-04-16 21:52:12,289 INFO Zookeeper connection established, continue actions.

**>>>>> Connection to zk reestablished <<<<<**

**>>>>> Semaphore checks if we have lease (we havent) <<<<<**
> 2017-04-16 21:52:12,808 DEBUG Sending request(xid=1): Exists(path=u/some_node/shard-semaphore/9e5849446fbe4fa18a2e18b7098c9a95, watcher=None)

**>>>>> Lock ensures path  <<<<<**
> 2017-04-16 21:52:16,847 DEBUG Sending request(xid=2): Exists(path=u/some_node/shard-semaphore-__lock__, watcher=None)
> 2017-04-16 21:52:20,672 DEBUG Received response(xid=2): ZnodeStat(czxid=21477875405, mzxid=21477875405, ctime=1492368653082, mtime=1492368653082, version=0, cversion=9, aversion=0, ephemeralOwner=0, dataLength=0, numChildren=3, pzxid=21477896863)

**>>>>> Lock creates sequence ephemeral node  <<<<<**
> 2017-04-16 21:52:20,672 DEBUG Sending request(xid=3): Create(path=u/some_node/shard-semaphore-__lock__/824c95c7e3f849f7a75d414ac78941dd__lock__, data=xxx.yyy.net:28137, acl=[ACL(perms=31, acl_list=[ALL], id=Id(scheme=world, id=anyone))], flags=3)
> 2017-04-16 21:52:22,603 DEBUG Received response(xid=3): u/some_node/shard-semaphore-__lock__/824c95c7e3f849f7a75d414ac78941dd__lock__0000000006

**>>>>> Lock gets children, sort them and check if his node has least sequence number <<<<<**
> 2017-04-16 21:52:22,603 DEBUG Sending request(xid=4): GetChildren(path=u/some_node/shard-semaphore-__lock__, watcher=None)
> 2017-04-16 21:52:24,853 DEBUG Received response(xid=4): [u824c95c7e3f849f7a75d414ac78941dd__lock__0000000006, ud981dca6219046498f544c052770652b__lock__0000000002, ub692b6d2d482453b95feb69c68566be6__lock__0000000005, ue9eddb1df31949a59b5b7a49312367c4__lock__0000000003]

**>>>>> Zombie node ud981dca6219046498f544c052770652b__lock__0000000002 created by same instance on previous iteration causes deadlock, delete lock lock node try again and forever <<<<<**
> 2017-04-16 21:52:24,853 DEBUG Sending request(xid=5): Delete(path=u/some_node/shard-semaphore-__lock__/824c95c7e3f849f7a75d414ac78941dd__lock__0000000006, version=-1)
> 2017-04-16 21:52:27,350 DEBUG Received response(xid=5): True

The obvious fix is to move `try` in  kazoo.recipe.lock.Semaphore._inner_acquire upper, so release in finally will be called and zombie lock will be deleted:
```
diff --git a/kazoo/recipe/lock.py b/kazoo/recipe/lock.py
index d351539..91c006c 100644
--- a/kazoo/recipe/lock.py
+++ b/kazoo/recipe/lock.py
@@ -472,10 +472,10 @@ class Semaphore(object):
         w = _Watch(duration=timeout)
         w.start()
         lock = self.client.Lock(self.lock_path, self.data)
-        gotten = lock.acquire(blocking=blocking, timeout=w.leftover())
-        if not gotten:
-            return False
-        try:
+        try
+            gotten = lock.acquire(blocking=blocking, timeout=w.leftover())
+            if not gotten:
+                return False
             while True:
                 self.wake_event.clear()
```

Am I right? Should I prepare pull request with fix?

**Comment by bbangert** [comment]
@anskornyakov Hi, thanks for the PR and filing this. I should probably remove myself as the maintainer of this, as I don't have the time to do a proper job of it. Unfortunately with something this complex, it can take me several days of study to come fully back up to speed on the intricacies and nuance of the system and the code. I no longer have such time available, so I don't feel qualified at the moment to indicate whether this is a proper solution or not.

There's a backlog of PR's, and I'm attempting to determine a proper solution, as I'd like to see kazoo get this and other important fixes that feel like they've been stalled in the backlog.

Have you tested this in your production system? It passes the integration tests which is great, if you have a report of how it's done in your system, I'll merge it based on that.

**Comment by bbangert** [comment]
Closed by #430.
